NorthwindCRUD/Services/PagingService.cs

```diff
@@ -21,38 +26,55 @@ public PagingService(IMapper mapper)
             this.mapper = mapper;
         }
 
-        public PagedResultDto<TDto> GetPagedData<TEntity, TDto>(IEnumerable<TEntity> data, int? skip, int? top, string? orderBy)
+        public PagedResultDto<TDto> FetchPagedData<TEntity, TDto>(
+            IQueryable<TEntity> query,
+            int? skip = null,
+            int? top = null,
+            int? pageIndex = null,
+            int? size = null,
+            string? orderBy = null)
         {
-            var totalRecords = data.Count();
+            // Determine if we're using skip/top or page/size
+            bool isPageSize = pageIndex.HasValue || size.HasValue;
 
-            // Default skip and top if not provided
+            int totalRecords = query.Count();
+
+            // Default values if not provided
             int skipRecordsAmount = skip ?? 0;
             int currentSize = top ?? totalRecords;
 
+            if (isPageSize)
+            {
+                int pageNumber = pageIndex ?? 0;
+                int pageSize = size ?? totalRecords;
+                skipRecordsAmount = pageNumber * pageSize;
+                currentSize = pageSize;
+            }
+
             // Apply ordering if specified
             if (!string.IsNullOrEmpty(orderBy))
             {
-                data = ApplyOrdering(data.AsQueryable(), orderBy).ToArray();
+                query = ApplyOrdering(query, orderBy);
             }
 
             // Apply pagination
-            var pagedData = data
+            var pagedData = query
                 .Skip(skipRecordsAmount)
                 .Take(currentSize)
                 .ToArray();
 
             // Calculate total pages
             int totalPages = (int)Math.Ceiling(totalRecords / (double)currentSize);
 
-            // Map the results to ProductDto
+            // Map the results to TDto
             var pagedDataDtos = mapper.Map<TDto[]>(pagedData);
 
             return new PagedResultDto<TDto>
             {
                 Items = pagedDataDtos,
                 TotalRecordsCount = totalRecords,
-                PageSize = pagedDataDtos.Length,
-                PageNumber = (skipRecordsAmount / currentSize) + 1,
+                PageSize = currentSize,
+                PageNumber = isPageSize ? ((skipRecordsAmount / currentSize) + 1) : ((skipRecordsAmount / currentSize) + 1),
```

Comment:
Can be improved like so ? (skipRecordsAmount / currentSize) : (skipRecordsAmount / currentSize) + 1